File: rust/src/unsafe/cell.rs
pub struct Cell {
    pub x: u32,
    pub y: u32,
    pub alive: bool,
    pub next_state: bool,
    pub neighbours: Vec<*const Cell>,
}

impl Cell {
    pub fn new(x: u32, y: u32, alive: bool) -> Self {
        Self {
            x,
            y,
            alive,
            next_state: false,
            neighbours: Vec::new(),
        }
    }

    pub fn to_char(&self) -> char {
        if self.alive { 'o' } else { ' ' }
    }

    pub fn alive_neighbours(&self) -> u32 {
        // The following is slower
        // self.neighbours
        //     .iter()
        //     .copied()
        //     .filter(|ptr| unsafe { (**ptr).alive })
        //     .count()

        // The following is the fastest
        let mut alive_neighbours = 0;
        for ptr in self.neighbours.iter().copied() {
            unsafe {
                if (*ptr).alive {
                    alive_neighbours += 1;
                }
            }
        }
        alive_neighbours

        // The following is about the same speed
        // let mut alive_neighbours = 0;
        // let count = self.neighbours.len();
        // for i in 0..count {
        //     let ptr = self.neighbours[i];
        //     unsafe {
        //         if (*ptr).alive {
        //             alive_neighbours += 1;
        //         }
        //     }
        // }
        // alive_neighbours
    }
}

File: objective-c/Cell.m
#import "Cell.h"

@implementation Cell

- (instancetype)initWithX:(NSUInteger)x y:(NSUInteger)y alive:(BOOL)alive {
  self = [super init];
  if (self) {
    _x = x;
    _y = y;
    _alive = alive;
    _nextState = nil;
    _neighbours = [NSMutableArray array];
  }
  return self;
}

- (NSString *)toChar {
  return self.alive ? @"o" : @" ";
}

- (NSUInteger)aliveNeighbours {
  // The following is slower
  // NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(Cell *cell, NSDictionary *bindings) {
  //   return cell.alive;
  // }];
  // return [[self.neighbours filteredArrayUsingPredicate:predicate] count];

  // The following is the fastest
  NSUInteger aliveNeighbours = 0;
  for (Cell *neighbour in self.neighbours) {
    if (neighbour.alive) {
      aliveNeighbours++;
    }
  }
  return aliveNeighbours;

  // The following is slower
  // NSUInteger aliveNeighbours = 0;
  // NSUInteger count = [self.neighbours count];
  // for (NSUInteger i = 0; i < count; i++) {
  //   Cell *neighbour = self.neighbours[i];
  //   if (neighbour.alive) {
  //     aliveNeighbours++;
  //   }
  // }
  // return aliveNeighbours;
}

@end
